File: src/swe_forge/execution/resources.py
"""Resource limits and cleanup policies for Docker container execution.

This module provides execution resource limits based on difficulty levels
and cleanup policies for container lifecycle management.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    pass

# =============================================================================
# Constants
# =============================================================================

#: Exit code indicating OOM kill (SIGKILL, typically from OOM or manual kill)
OOM_EXIT_CODE = 137

#: Default memory limit in MB
DEFAULT_MEMORY_MB = 512

#: Default CPU limit (1.0 = 1 full CPU core)
DEFAULT_CPU_CORES = 1.0

#: Default disk space limit in GB
DEFAULT_DISK_GB = 5

#: Default maximum processes
DEFAULT_MAX_PROCESSES = 100

#: Default timeout in seconds (20 minutes)
DEFAULT_TIMEOUT_SECONDS = 1200

#: CPU period in microseconds (100ms is the standard period)
CPU_PERIOD_US = 100_000


# =============================================================================
# Enums
# =============================================================================


class CleanupPolicy(Enum):
    """Policy for container cleanup after execution.

    Attributes:
        ALWAYS: Always remove containers after execution (default).
        ON_FAILURE: Only remove containers when execution fails.
        NEVER: Never automatically remove containers.
    """

    ALWAYS = "always"
    ON_FAILURE = "on_failure"
    NEVER = "never"


# =============================================================================
# Dataclasses
# =============================================================================


@dataclass
class ResourceLimits:
    """Execution resource limits for a container.

    These limits control the resources available to a container
    during task execution.

    Attributes:
        memory_mb: Memory limit in megabytes.
        cpu_cores: CPU cores available (e.g., 0.5, 1.0, 2.0).
        disk_gb: Disk space limit in gigabytes.
        max_processes: Maximum number of processes allowed (PIDs limit).
        timeout_seconds: Timeout in seconds before the container is killed.
    """

    memory_mb: int = DEFAULT_MEMORY_MB
    cpu_cores: float = DEFAULT_CPU_CORES
    disk_gb: int = DEFAULT_DISK_GB
    max_processes: int = DEFAULT_MAX_PROCESSES
    timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS

    def memory_bytes(self) -> int:
        """Return memory limit in bytes.

        Returns:
            Memory limit as number of bytes.
        """
        return self.memory_mb * 1024 * 1024

    def cpu_period(self) -> int:
        """Return CPU period in microseconds.

        Uses the standard 100ms (100,000 microseconds) period.

        Returns:
            CPU scheduler period in microseconds.
        """
        return CPU_PERIOD_US

    def cpu_quota(self) -> int:
        """Return CPU quota in microseconds.

        The quota is calculated as: quota = period * cores
        For example, 1.0 core = 100,000 quota (100% of one CPU)

        Returns:
            CPU quota in microseconds.
        """
        return int(self.cpu_period() * self.cpu_cores)

    def disk_bytes(self) -> int:
        """Return disk space limit in bytes.

        Returns:
            Disk limit as number of bytes.
        """
        return self.disk_gb * 1024 * 1024 * 1024

    def nano_cpus(self) -> int:
        """Return CPU limit in nanoseconds (NanoCpus format).

        Docker's NanoCpus is 1 billion nanoseconds per CPU.
        For example, 1.0 CPU = 1,000,000,000 nanoseconds.

        Returns:
            CPU limit in nanoseconds.
        """
        return int(self.cpu_cores * 1_000_000_000)


# =============================================================================
# Helper Functions
# =============================================================================


def is_oom_killed(exit_code: int | None) -> bool:
    """Check if the exit code indicates an OOM kill.

    Exit code 137 indicates the container was killed with SIGKILL,
    which typically happens due to OOM or a manual kill signal.

    Args:
        exit_code: The exit code from container execution.

    Returns:
        True if the exit code indicates OOM kill.

    Example:
        >>> is_oom_killed(137)
        True
        >>> is_oom_killed(0)
        False
    """
    return exit_code == OOM_EXIT_CODE


def apply_resource_limits(
    container_config: dict[str, Any],
    limits: ResourceLimits,
) -> dict[str, Any]:
    """Apply resource limits to a Docker container configuration.

    Modifies the container config's HostConfig to include resource limits.
    Creates HostConfig if it doesn't exist.

    Args:
        container_config: Docker container configuration dict to modify.
        limits: ResourceLimits instance with the limits to apply.

    Returns:
        The modified container configuration dict.

    Example:
        >>> config = {"Image": "python:3.11"}
        >>> limits = ResourceLimits(memory_mb=1024, cpu_cores=2.0)
        >>> apply_resource_limits(config, limits)
        {'Image': 'python:3.11', 'HostConfig': {'Memory': 1073741824, ...}}
    """
    if "HostConfig" not in container_config:
        container_config["HostConfig"] = {}

    host_config = container_config["HostConfig"]

    # Memory limit in bytes
    host_config["Memory"] = limits.memory_bytes()

    # Disable swap limit (set to -1 to use same as memory)
    host_config["MemorySwap"] = -1

    # CPU limits
    # Using NanoCpus for Docker API compatibility
    host_config["NanoCpus"] = limits.nano_cpus()

    # CPU quota/period (alternative method, also supported)
    host_config["CpuPeriod"] = limits.cpu_period()
    host_config["CpuQuota"] = limits.cpu_quota()

    # Process limit (PIDs)
    host_config["PidsLimit"] = limits.max_processes

    return container_config


def get_resource_limits(difficulty: str) -> ResourceLimits:
    """Get execution limits based on difficulty level.

    Supported difficulty levels (case-insensitive):
        - "easy": Light resources for simple tasks
        - "medium": Moderate resources for standard tasks (default)
        - "hard": Heavy resources for complex tasks
        - "expert": Very high resources for expert-level tasks
        - "nightmare": Maximum resources for extreme tasks

    Unknown difficulty levels default to "medium".

    Args:
        difficulty: The difficulty level as a string (case-insensitive).

    Returns:
        ResourceLimits appropriate for the difficulty level.

    Example:
        >>> limits = get_resource_limits("hard")
        >>> limits.memory_mb
        2048
        >>> limits.cpu_cores
        2.0
    """
    difficulty_lower = difficulty.lower()

    if difficulty_lower == "easy":
        return ResourceLimits(
            memory_mb=512,
            cpu_cores=0.5,
            disk_gb=2,
            max_processes=50,
            timeout_seconds=600,  # 10 minutes
        )

    if difficulty_lower == "medium":
        return ResourceLimits(
            memory_mb=1024,
            cpu_cores=1.0,
            disk_gb=5,
            max_processes=100,
            timeout_seconds=1200,  # 20 minutes
        )

    if difficulty_lower == "hard":
        return ResourceLimits(
            memory_mb=2048,
            cpu_cores=2.0,
            disk_gb=10,
            max_processes=200,
            timeout_seconds=2400,  # 40 minutes
        )

    if difficulty_lower == "expert":
        return ResourceLimits(
            memory_mb=4096,
            cpu_cores=4.0,
            disk_gb=20,
            max_processes=500,
            timeout_seconds=4800,  # 80 minutes
        )

    if difficulty_lower == "nightmare":
        return ResourceLimits(
            memory_mb=8192,
            cpu_cores=8.0,
            disk_gb=50,
            max_processes=1000,
            timeout_seconds=9000,  # 150 minutes
        )

    # Unknown difficulty defaults to medium
    return ResourceLimits(
        memory_mb=1024,
        cpu_cores=1.0,
        disk_gb=5,
        max_processes=100,
        timeout_seconds=1200,
    )


def should_cleanup(policy: CleanupPolicy, success: bool) -> bool:
    """Determine if cleanup should occur based on policy and execution result.

    Args:
        policy: The cleanup policy to apply.
        success: Whether the execution was successful.

    Returns:
        True if cleanup should be performed.

    Example:
        >>> should_cleanup(CleanupPolicy.ALWAYS, True)
        True
        >>> should_cleanup(CleanupPolicy.ON_FAILURE, True)
        False
        >>> should_cleanup(CleanupPolicy.NEVER, False)
        False
    """
    if policy == CleanupPolicy.ALWAYS:
        return True

    if policy == CleanupPolicy.ON_FAILURE:
        return not success

    # CleanupPolicy.NEVER
    return False

File: src/execution/resources.rs
//! Resource limits for Docker container execution.
//!
//! This module provides execution resource limits based on difficulty levels,
//! mirroring the existing difficulty system in the codebase.

use serde::{Deserialize, Serialize};

/// Execution resource limits for a container.
///
/// These limits control the resources available to a container
/// during task execution.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExecutionLimits {
    /// Memory limit in megabytes.
    pub memory_mb: u64,
    /// CPU cores available (e.g., 0.5, 1.0, 2.0).
    pub cpu_cores: f64,
    /// Disk space limit in gigabytes.
    pub disk_gb: u64,
    /// Maximum number of processes allowed.
    pub max_processes: u64,
    /// Timeout in seconds before the container is killed.
    pub timeout_seconds: u64,
}

impl Default for ExecutionLimits {
    fn default() -> Self {
        // Default to medium difficulty limits
        Self {
            memory_mb: 1024,
            cpu_cores: 1.0,
            disk_gb: 5,
            max_processes: 100,
            timeout_seconds: 1200, // 20 minutes
        }
    }
}

impl ExecutionLimits {
    /// Creates new execution limits with the given parameters.
    pub fn new(
        memory_mb: u64,
        cpu_cores: f64,
        disk_gb: u64,
        max_processes: u64,
        timeout_seconds: u64,
    ) -> Self {
        Self {
            memory_mb,
            cpu_cores,
            disk_gb,
            max_processes,
            timeout_seconds,
        }
    }

    /// Returns memory limit in bytes.
    pub fn memory_bytes(&self) -> i64 {
        (self.memory_mb * 1024 * 1024) as i64
    }

    /// Returns CPU period in nanoseconds (fixed at 100ms).
    pub fn cpu_period(&self) -> i64 {
        100_000 // 100ms in microseconds
    }

    /// Returns CPU quota based on cores allocated.
    ///
    /// Formula: quota = period * cores
    /// e.g., 1.0 core = 100000 quota (100% of one CPU)
    pub fn cpu_quota(&self) -> i64 {
        (self.cpu_period() as f64 * self.cpu_cores) as i64
    }

    /// Returns disk space limit in bytes.
    pub fn disk_bytes(&self) -> u64 {
        self.disk_gb * 1024 * 1024 * 1024
    }
}

/// Get execution limits based on difficulty level string.
///
/// Supported difficulty levels:
/// - "easy" - Light resources for simple tasks
/// - "medium" - Moderate resources for standard tasks  
/// - "hard" - Heavy resources for complex tasks
/// - "expert" - Very high resources for expert-level tasks
/// - "nightmare" - Maximum resources for extreme tasks
///
/// Unknown difficulty levels default to "medium".
///
/// # Arguments
///
/// * `difficulty` - The difficulty level as a string (case-insensitive)
///
/// # Returns
///
/// Execution limits appropriate for the difficulty level.
///
/// # Example
///
/// ```
/// use swe_forge::execution::get_execution_limits;
///
/// let limits = get_execution_limits("hard");
/// assert_eq!(limits.memory_mb, 2048);
/// assert_eq!(limits.timeout_seconds, 2400);
/// ```
pub fn get_execution_limits(difficulty: &str) -> ExecutionLimits {
    match difficulty.to_lowercase().as_str() {
        "easy" => ExecutionLimits {
            memory_mb: 512,
            cpu_cores: 0.5,
            disk_gb: 2,
            max_processes: 50,
            timeout_seconds: 600, // 10 minutes
        },
        "medium" => ExecutionLimits {
            memory_mb: 1024,
            cpu_cores: 1.0,
            disk_gb: 5,
            max_processes: 100,
            timeout_seconds: 1200, // 20 minutes
        },
        "hard" => ExecutionLimits {
            memory_mb: 2048,
            cpu_cores: 2.0,
            disk_gb: 10,
            max_processes: 200,
            timeout_seconds: 2400, // 40 minutes
        },
        "expert" => ExecutionLimits {
            memory_mb: 4096,
            cpu_cores: 4.0,
            disk_gb: 20,
            max_processes: 500,
            timeout_seconds: 4800, // 80 minutes
        },
        "nightmare" => ExecutionLimits {
            memory_mb: 8192,
            cpu_cores: 8.0,
            disk_gb: 50,
            max_processes: 1000,
            timeout_seconds: 9000, // 150 minutes
        },
        // Unknown difficulty defaults to medium
        _ => ExecutionLimits::default(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_easy_limits() {
        let limits = get_execution_limits("easy");
        assert_eq!(limits.memory_mb, 512);
        assert_eq!(limits.cpu_cores, 0.5);
        assert_eq!(limits.disk_gb, 2);
        assert_eq!(limits.max_processes, 50);
        assert_eq!(limits.timeout_seconds, 600);
    }

    #[test]
    fn test_medium_limits() {
        let limits = get_execution_limits("medium");
        assert_eq!(limits.memory_mb, 1024);
        assert_eq!(limits.cpu_cores, 1.0);
        assert_eq!(limits.disk_gb, 5);
        assert_eq!(limits.max_processes, 100);
        assert_eq!(limits.timeout_seconds, 1200);
    }

    #[test]
    fn test_hard_limits() {
        let limits = get_execution_limits("hard");
        assert_eq!(limits.memory_mb, 2048);
        assert_eq!(limits.cpu_cores, 2.0);
        assert_eq!(limits.disk_gb, 10);
        assert_eq!(limits.max_processes, 200);
        assert_eq!(limits.timeout_seconds, 2400);
    }

    #[test]
    fn test_expert_limits() {
        let limits = get_execution_limits("expert");
        assert_eq!(limits.memory_mb, 4096);
        assert_eq!(limits.cpu_cores, 4.0);
        assert_eq!(limits.disk_gb, 20);
        assert_eq!(limits.max_processes, 500);
        assert_eq!(limits.timeout_seconds, 4800);
    }

    #[test]
    fn test_nightmare_limits() {
        let limits = get_execution_limits("nightmare");
        assert_eq!(limits.memory_mb, 8192);
        assert_eq!(limits.cpu_cores, 8.0);
        assert_eq!(limits.disk_gb, 50);
        assert_eq!(limits.max_processes, 1000);
        assert_eq!(limits.timeout_seconds, 9000);
    }

    #[test]
    fn test_case_insensitive() {
        let limits_lower = get_execution_limits("hard");
        let limits_upper = get_execution_limits("HARD");
        let limits_mixed = get_execution_limits("Hard");

        assert_eq!(limits_lower.memory_mb, limits_upper.memory_mb);
        assert_eq!(limits_lower.memory_mb, limits_mixed.memory_mb);
    }

    #[test]
    fn test_unknown_defaults_to_medium() {
        let limits = get_execution_limits("unknown");
        let medium = get_execution_limits("medium");

        assert_eq!(limits.memory_mb, medium.memory_mb);
        assert_eq!(limits.cpu_cores, medium.cpu_cores);
        assert_eq!(limits.timeout_seconds, medium.timeout_seconds);
    }

    #[test]
    fn test_memory_bytes_conversion() {
        let limits = ExecutionLimits::new(512, 1.0, 5, 100, 300);
        assert_eq!(limits.memory_bytes(), 512 * 1024 * 1024);
    }

    #[test]
    fn test_cpu_quota_calculation() {
        let limits = ExecutionLimits::new(512, 2.0, 5, 100, 300);
        assert_eq!(limits.cpu_period(), 100_000);
        assert_eq!(limits.cpu_quota(), 200_000); // 2 cores = 2 * period
    }

    #[test]
    fn test_disk_bytes_conversion() {
        let limits = ExecutionLimits::new(512, 1.0, 10, 100, 300);
        assert_eq!(limits.disk_bytes(), 10 * 1024 * 1024 * 1024);
    }
}
